controller/contour/service.go

```diff
@@ -0,0 +1,255 @@
+	// [TODO] danehans: Update Contour name to contour.Name + "-contour" to support multiple
+	// Contours/ns when https://github.com/projectcontour/contour/issues/2122 is fixed.
+	contourSvcName = "contour"
+	// [TODO] danehans: Update Envoy name to contour.Name + "-envoy" to support multiple
+	// Contours/ns when https://github.com/projectcontour/contour/issues/2122 is fixed.
+	// envoySvcName is the name of Envoy's Service.
+	envoySvcName = "envoy"
+	// awsLbBackendProtoAnnotation is a Service annotation that places the AWS ELB into
+	// "TCP" mode so that it does not do HTTP negotiation for HTTPS connections at the
+	// ELB edge. The downside of this is the remote IP address of all connections will
+	// appear to be the internal address of the ELB. See docs/proxy-proto.md for information
+	// about enabling the PROXY protocol on the ELB to recover the original remote IP address.
+	// TODO [danehans]: Make proxy protocol configurable or automatically enabled. See
+	// https://github.com/projectcontour/contour-operator/issues/49 for details.
+	awsLbBackendProtoAnnotation = "service.beta.kubernetes.io/aws-load-balancer-backend-protocol"
+)
+
+// ensureContourService ensures that a Contour Service exists for the given contour.
+func (r *Reconciler) ensureContourService(ctx context.Context, contour *operatorv1alpha1.Contour) error {
+	desired := DesiredContourService(contour)
+
+	current, err := r.currentContourService(ctx, contour)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return r.createService(ctx, desired)
+		}
+		return fmt.Errorf("failed to get service %s/%s: %w", desired.Namespace, desired.Name, err)
+	}
+
+	if err := r.updateContourServiceIfNeeded(ctx, current, desired); err != nil {
+		return fmt.Errorf("failed to update service %s/%s: %w", desired.Namespace, desired.Name, err)
+	}
+
+	return nil
+}
+
+// ensureEnvoyService ensures that an Envoy Service exists for the given contour.
+func (r *Reconciler) ensureEnvoyService(ctx context.Context, contour *operatorv1alpha1.Contour) error {
+	desired := DesiredEnvoyService(contour)
+
+	current, err := r.currentEnvoyService(ctx, contour)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return r.createService(ctx, desired)
+		}
+		return fmt.Errorf("failed to get service %s/%s: %w", desired.Namespace, desired.Name, err)
+	}
+
+	if err := r.updateEnvoyServiceIfNeeded(ctx, current, desired); err != nil {
+		return fmt.Errorf("failed to update service %s/%s: %w", desired.Namespace, desired.Name, err)
+	}
+
+	return nil
+}
+
+// ensureContourServiceDeleted ensures that a Contour Service for the
+// provided contour is deleted.
+func (r *Reconciler) ensureContourServiceDeleted(ctx context.Context, contour *operatorv1alpha1.Contour) error {
+	svc := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: contour.Spec.Namespace.Name,
+			Name:      contourSvcName,
+		},
+	}
+
+	if err := r.Client.Delete(ctx, svc); err != nil {
+		if !errors.IsNotFound(err) {
+			return err
```

Comment:
Might be worth wrapping errors here and in `ensureEnvoyServiceDeleted` in case the error message doesn't already indicate the name of the service:
```suggestion
			return fmt.Errorf("failed to delete service %s/%s: %w", svc.Namespace, svc.Name, err)
```